src/Microsoft.SymbolManifestGenerator/SymbolManifestGenerator.cs

```diff
@@ -0,0 +1,112 @@
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using Microsoft.SymbolStore;
+using Microsoft.SymbolStore.KeyGenerators;
+
+namespace Microsoft.SymbolManifestGenerator
+{
+    public class SymbolManifestGenerator
+    {
+        private const int Private = 340;
+
+        public static void GenerateManifest(ITracer tracer, DirectoryInfo dir, string manifestFileName)
+        {
+            ManifestDataV1 manifestData = new();
+
+            foreach (FileInfo file in dir.GetFiles("*", SearchOption.AllDirectories))
+            {
+                using FileStream fileStream = file.OpenRead();
+                SymbolStoreFile symbolStoreFile = new(fileStream, file.FullName);
+                FileKeyGenerator generator = new(tracer, symbolStoreFile);
+                if (!generator.IsValid())
+                {
+                    tracer.Information($"Could not generate a valid FileKeyGenerator from file '{file.FullName}'. Skipping.");
+                    continue;
+                }
+
+                foreach (SymbolStoreKey clrKey in generator.GetKeys(KeyTypeFlags.ClrKeys))
+                {
+                    string runtimeModuleDirectory = file.DirectoryName;
+                    FileInfo specialFile = GetSymbolFileToAddAdditionalDebugEntry(runtimeModuleDirectory, clrKey);
+                    if (specialFile == null)
+                    {
+                        tracer.Information($"Known special file '{clrKey.FullPathName}' for runtime module '{file.FullName}' does not exist in directory '{runtimeModuleDirectory}'. Skipping.");
+                        continue;
+                    }
+
+                    string basedirRelativePath = Path.Combine(runtimeModuleDirectory.Replace(dir.FullName, string.Empty).TrimStart(Path.DirectorySeparatorChar), specialFile.Name);
+                    string fileHash = CalculateSHA512(specialFile);
+
+                    ManifestDataEntry manifestDataEntry = new()
+                    {
+                        basedirRelativePath = basedirRelativePath,
+                        symbolKey = clrKey.Index,
+                        sha512 = fileHash,
+                        debugInformationLevel = Private
+                    };
+
+                    manifestData.entries.Add(manifestDataEntry);
+                }
+            }
+
+            string manifestDataContent = JsonSerializer.Serialize(manifestData);
+            File.WriteAllText(manifestFileName, manifestDataContent);
+        }
+
+        private static FileInfo GetSymbolFileToAddAdditionalDebugEntry(string runtimeModuleDir, SymbolStoreKey clrKey)
+        {
+            string keyFileNameToMatch = clrKey.FullPathName;
+
+            string fileOnDisk = Directory.EnumerateFiles(runtimeModuleDir)
+                                      .FirstOrDefault(f => Path.GetFileName(f).Equals(keyFileNameToMatch, StringComparison.OrdinalIgnoreCase));
```

Comment:
We should use First. Essentially - either we only publish one or fail.

Reply:
`First` will throw if there is no match (i.e. the known special file isn't found on disk). Default allows us to gracefully continue in a situation where not all known ClrKey's from a runtime module are found in the basedir, which wouldn't be uploaded in the first place.

I think if we use `First` instead of `FirstOrDefault`, that means we are requiring that all known special files are being uploaded alongside a runtime module, rather than just what are found. Do we want that constraint?
(e.g. [`coreclr.dll`, `mscordaccore.dll`, `mscordbi.dll`] would throw because there are no long name dac variants, [`coreclr.dll`, `mscordaccore.dll`, `mscordaccore_long_name_variants.dll`s] would throw because there is no `mscordbi.dll`
only [`coreclr.dll`, `mscordaccore.dll`, `mscordaccore_long_name_variants.dll`s, `mscordbi.dll`]  would not throw.)

That constraint would prevent any manifest generation for a subset of special files, say we wanted to reupload `mscordbi.dll` under the clrkey, we would have to give this assembly a directory that contains all `coreclr.dll` special files instead of just [`coreclr.dll`, `mscordbi.dll`]

Reply:
I think if we want to instead counter situations where we have multiple file matches, we should use `SingleOrDefault`